presto-main/src/main/java/com/facebook/presto/operator/exchange/AsyncWaiter.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+
+class AsyncWaiter
```

Comment:
The guarantee provided by this class does not seem enough for the use case.

Let's say, a thread is inside `waitForWriting`:
```
        synchronized (this) {
            if (bufferBytes.get() < maxBufferedBytes) {
                return NOT_BLOCKED;
            }
        }
        return writeWaiter.waitFor();
```
The thread has finished the sync block inside the function, but hasn't executed `writerWaiter.waitFor`.
Meanwhile, another thread executes and finishes
```
            long bytes = bufferBytes.addAndGet(delta);
            if (delta < 0 && bytes < maxBufferedBytes) {
                writeWaiter.notifyWaiters();
            }
```
As a result, the first thread can become blocked when there is actually enough empty slots to write to. It will be unblocked only after somehow the buffer becomes full and then becomes not full. But this may never happen if enough threads gets into the same situation as the first thread.